### PR TITLE
Enabling TDM topology and supporting simultaneous playback and capture usecases

### PR DESCRIPTION
**Pull request by maruthi-machani**

### CODE REVIEW

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 1]
```diff
@@ -98,7 +98,7 @@ DAI_CONFIG(ACPHS, 1, 1, acp-amp-codec,
 # PCM Playback pipeline 4 on PCM 1 using max 2 channels of s16le.
 # 10ms deadline with priority 0 on core 0
 # this is connected to pipeline DAI 1
-PIPELINE_PCM_ADD(sof/pipe-host-playback.m4,
+PIPELINE_PCM_ADD(sof/pipe-passthrough-playback-sched.m4,
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:101]: This seems inconsistent with the description of pipelines

````
# PCM0 ----> buffer ------+
#                         |----> muxdemux ----> buffer ---->  ACPSP_play
# PCM1 ----> buffer ----->
````

It's very hard to review if you have the m4 stuff without updating what the topology looks like...

File: tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4 [hunk 2]
```diff
@@ -0,0 +1,57 @@
+# Low Latency Passthrough Pipeline and PCM
+#
+# Pipeline Endpoints for connection are :-
+#
+#  host PCM_P --> B0 --> sink DAI0
```
Comment (plbossart) [inline tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4:5]: this clearly isn't aligned with the use of a MUX?

**Review by plbossart — COMMENTED** [review]
Very confusing PR, completely unclear how mux and passthrough concept can coexist ...

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 3]
```diff
@@ -64,8 +68,8 @@ define(matrix2, `ROUTE_MATRIX(4,
 dnl name, num_streams, route_matrix list
 MUXDEMUX_CONFIG(demux_priv_1, 2, LIST_NONEWLINE(`', `matrix1,', `matrix2'))
 
-# playback DAI is ACPSP using 2 periods
-# Buffers use s16le format, 1000us deadline on core 0 with priority 1
+# playback DAI is ACPHS using 2 periods
+# Buffers use s16le format, 2000us deadline on core 0 with priority 1
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:72]: don't mix comment changes with functionality changes, this has to go in a separate patch

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 4]
```diff
@@ -96,9 +100,9 @@ DAI_CONFIG(ACPHS, 1, 1, acp-amp-codec,
                 ACP_TDM(8, 32, 3, 3),ACPHS_CONFIG_DATA(ACPHS, 1, 48000, 8, 1)))
 #/**********************************************************************************/
 # PCM Playback pipeline 4 on PCM 1 using max 2 channels of s16le.
-# 10ms deadline with priority 0 on core 0
+# 2000us deadline with priority 0 on core 0
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:103]: separate patch

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 5]
```diff
@@ -141,8 +156,8 @@ PIPELINE_PCM_ADD(sof/pipe-low-latency-capture-demux.m4,
         2000, 0, 0,
         48000, 48000, 48000)
 
-# Capture DAI is ACPSP using 2 periods
-# Buffers use s16le format, 1000us deadline with priority 0 on core 0
+# Capture DAI is ACPHS using 2 periods
+# Buffers use s16le format, 2000us deadline with priority 0 on core 0
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:160]: separate patch

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4 [hunk 6]
```diff
@@ -65,7 +68,7 @@ define(matrix2, `ROUTE_MATRIX(4,
 dnl name, num_streams, route_matrix list
 MUXDEMUX_CONFIG(demux_priv_1, 2, LIST_NONEWLINE(`', `matrix1,', `matrix2'))
 
-# playback DAI is ACPSP using 2 periods
+# playback DAI is ACPHS using 2 periods
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:71]: separate patch

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 7]
```diff
@@ -1,5 +1,5 @@
 #
-# Topology for Renoir with I2S SP and DMIC.
+# Topology for Rembrandt with I2S HS and DMIC.
```
Comment (plbossart) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:2]: separate patch

**Review by plbossart — CHANGES_REQUESTED** [review]
please don't mix cosmetic changes or corrections with functionality changes, it makes the reviews impossible.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 3 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:72]: created seperate patch for comments

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 5 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:160]: created seperate patch for comments

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4 (hunk 6 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:71]: created seperate patch for comments

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 7 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:2]: created seperate patch for comments

**Comment by maruthi-machani** [comment]
created separate patch for cosmetic changes.

**Comment by maruthi-machani** [comment]
@plbossart  any further review comments or can be merged?

**Review by plbossart — COMMENTED** [review]
I fundamentally don't understand the topology and after multiple reviews I'll have to pass on providing further comments.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 8]
```diff
@@ -32,13 +32,17 @@ VIRTUAL_WIDGET(ACPHS1 OUT, output, 3)
 #
 # Define the pipelines
 #
-# PCM0 ----> buffer ------+
-#                         |----> muxdemux ----> buffer ---->  ACPSP_play
-# PCM1 ----> buffer ----->
 #
-# PCM0 <---- muxdemux <---- buffer <---- ACPSP_Capture
+#              |-------------->echoref --->ACPHS1
```
Comment (cujomalainey) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:36]: im a little confused by this pictogram, it shows data coming from the host into the echoref? also im confused why its linked at the buffer and not a mux

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 9]
```diff
@@ -118,6 +122,17 @@ SectionGraph."PIPE_NAME" {
...
 PCM_PLAYBACK_ADD(Media Playback MUX 1, 1, PIPELINE_PCM_4)
 
+
+# Connect virtual capture to dai
+SectionGraph."PIPE_PLAYBACK_VIRT" {
+	index "1"
+
+	lines [
+		# mux to capture
+		dapm(ACPHS1.OUT, ECHO REF 4)
```
[the remark this answers is not in the document]
Comment (cujomalainey) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:132]: +1 as I would assume HS has nothing to do with echo ref

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4 [hunk 10]
```diff
@@ -33,13 +32,17 @@ VIRTUAL_WIDGET(ACPHS1 OUT, output, 3)
...
 # Define the pipelines
 #
-# PCM0 ----> buffer ------+
-#                         |----> muxdemux ----> buffer ---->  ACPSP_play
-# PCM1 ----> buffer ----->
 #
-# PCM0 <---- muxdemux <---- buffer <---- ACPSP_Capture
+#              |-------------->echoref --->ACPHS1_Playback
+#              |
+# PCM3 ----> buffer ------+
```
Comment (cujomalainey) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:38]: again very confused by linking at buffer

Comment (bhiregoudar) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:38]: I think it is the confusion with the name mentioned in the topology as it is recommended for reference from one of the Intel's topology which was used for the echo-ref. On AMD platform, it is intended to use for playback only. On AMD platform, we need a virtual/dummy DAI to support simultaneous playback on speaker and headset using single I2S controller instance, hence the connection of the buffer is to MUX and the reference buffer. So, connections to ACPHS1.OUT are kind of dummy only to get some of the required IPC messages from driver to enable two different PCM streams for headset and the speaker for simultaneous playback. It was the need of the design/framework, but improper naming confused.  We will update the patch with proper name and resubmit.  Let me know if any further queries.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhiregoudar) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:132]: @plbossart : As mentioned in the other comment, the intention is to support the simultaneous playback on headset and speaker with single I2S controller instance. So, there is a need of that type of topology. I understand that, there is confusion with the naming, will correct the naming.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4 (hunk 10 above)
Comment (marc-hb) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:38]: > We will update the patch with proper name and resubmit.

Some inline comments can't hurt either. When a relatively long explanation is needed on Github, some of it most likely belong to the corresponding source. It should not be required to use git or (even less convenient) github to understand what some source does.

My 2 cents.

**Comment by cujomalainey** [comment]
> Attaching the topology graph for reference
> 
> comparing the intel demux topology graph to amd mux topology graph for reference only. shows the virtual component(echo-ref) between the buffer and dai when a demux or mux component is used.

Both those graphs look incorrect

On intel we should not have two components writing to the same buf. I think the graph is missing an AEC which should consume a buffer between it and the DEMUX.

On AMD you have two reads of BUF4.0, also why do you have a echoref object in your HS out path?

**Comment by maruthi-machani** [comment]
> > Attaching the topology graph for reference
> > comparing the intel demux topology graph to amd mux topology graph for reference only. shows the virtual component(echo-ref) between the buffer and dai when a demux or mux component is used.
> 
> Both those graphs look incorrect
> 
> On intel we should not have two components writing to the same buf. I think the graph is missing an AEC which should consume a buffer between it and the DEMUX.
> 
> On AMD you have two reads of BUF4.0, also why do you have a echoref object in your HS out path?

The echo-ref is a virtual component without any functionality (echo-ref will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux). 

**Details of Topology:**
Three pipelines are used to support the mux use case and a separate path virtual component (echo-ref) to ACPHS1, this path has no functionality wrt sof FW.
two input host pipelines connected to the mux-dai(ACPHS0) pipeline.
image

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 8 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4]: The echo-ref is a virtual component without any functionality (echo-ref will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4 (hunk 10 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic4ch-rt5682-rt1019.m4:38]: The echo-ref is a virtual component without any functionality (echo-ref will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

**Comment by maruthi-machani** [comment]
@cujomalainey, @lgirdwood any further review comments or can be merged?

**Comment by bhiregoudar** [comment]
@lgirdwood @plbossart @cujomalainey : What is the next step to take this PR forward and merge?. To support simultaneous playback on speaker and headset with single I2S instance we need this kind of topology to be supported.

**Comment by bhiregoudar** [comment]
> I've commented multiple times that creating an echo reference for the sake of creating an echo reference, with no non-linear processing on the DSP, is completely useless. That creates a set of issues with pipeline management which could be completely avoided with a userspace loopback that would give the _same_ performance for AEC.

We have corrected it, we have changed it as virtual playback. We are not using any echo reference, it was a copy & paste mistake, which we have corrected now. We are only using the path/pipeline for playback purpose.  
If still some confusions are there due to the older comments, review points or pictures shared in this PR, then we will raise a new and fresh PR for you guys to look at it freshly. Let me know, if that is required.

**Comment by maruthi-machani** [comment]
Attaching the topology for reference, with virtual playback.

image

**Comment by plbossart** [comment]
I understand even less what the 'virtual playback' is.

**Comment by maruthi-machani** [comment]
> The virtual playback is a virtual component without any functionality (echo-ref will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

The echo-ref is a virtual component without any functionality (echo-ref will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

> I understand even less what the 'virtual playback' is.



> I understand even less what the 'virtual playback' is.



> I understand even less what the 'virtual playback' is.

The virtual playback is a virtual component without any functionality (virtual playback will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

> I understand even less what the 'virtual playback' is.

The virtual playback is a virtual component without any functionality (virtual playback will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

**Comment by maruthi-machani** [comment]
The virtual playback is a virtual component without any functionality (virtual playback will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).

**Comment by maruthi-machani** [comment]
The virtual playback is a virtual component without any functionality (virtual playback will not read/write from/to buffer), it is used only for termination of the ACPHS1 (2nd playback-dai of the mux).
For the driver we need to show two I2S endpoints are present(ACPHS and ACPHS1), but we do playback of two different streams through single I2S instance (ACPHS) due to hardware limitation, so virtual/dummy playback component is required to show two I2S endpoints presence in the system and support hardware params IPC commands.

**Comment by plbossart** [comment]
I am not following why 'termination of ACPHS1' is required.

**Comment by marc-hb** [comment]
Avoid screenshots of text please: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#use-copy-paste

**Comment by bhiregoudar** [comment]
> with current existing topology, seeing the below dmesg error, from the following commit [e8237f3](https://github.com/thesofproject/sof/commit/e8237f3e0b0c106983b3caa1aa4af417f08080a2). due to the improper termination of ACPHS1 dai.
> 
> image
> 
> Dmesg error: image
> 
> Commit details: image
> 
> This error is resolved with proper termination of ACPHS1(changes in this pull request)

As Maruthi mentioned, the issue is caused by the commit e8237f3e0b0c106983b3caa1aa4af417f08080a2, so we looked for an option to change the topology which resolved the issue caused by the commit e8237f3e0b0c106983b3caa1aa4af417f08080a2. So, we are expecting either to get the fix for the commit e8237f3e0b0c106983b3caa1aa4af417f08080a2 to solve the issue (Let us know if issue/ticket needs to be created) or need to go ahead with this PR.

File: tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4 [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+#
+# Pipeline Graph
+#
+#  host PCM_P --> B0 --> virtual-playback-passthrough
+
+P_GRAPH(pipe-passthrough-playback-sched, PIPELINE_ID,
+	LIST(`		',
+	`dapm(N_BUFFER(0), N_PCMP(PCM_ID))',
+	`dapm(VIRTUAL PLAYBACK PASSTHROUGH PIPELINE_ID, N_BUFFER(0))' ))
```
Comment (kv2019i) [inline tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4:39]: Considering this adds the "VIRTUAL PLAYBACK" dapm, the filename should probably have "virtual" in the name. I.e. this "pipe-passthrough-playback-sched.m4" hints this is more generic pipeline than it actually is.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 [hunk 12]
```diff
@@ -1,5 +1,5 @@
 #
```
Comment (kv2019i) [inline tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:1]: minor nit: Given this only changes "sof-acp*" files, a prefix like "topology1: acp: " would be nice.

**Review by kv2019i — COMMENTED** [review]
@bhiregou The explanation in comments helped.  It would help to have more of the context also in git commits to help the review. 

This is specially complex case, so extra verbosity will help. First, I understand you are working around an issue with TDM here. I don't fully understand why ACPHS1 needs to be present in topology (if no data is written to it), but I assume there's some dependency/need for this, and this has been in use before this PR. But this definitely is not a common topology construct.

Then the problem hit here is that SOF pipeline code assumes there's a single sink and source for a pipeline. Functions like ipc_get_ppl_sink_comp() and ipc_get_ppl_src_comp() are used in generic pipeline code and they assume a single component to be returned. You hit this with https://github.com/thesofproject/sof/commit/e8237f3e0b0c106983b3caa1aa4af417f08080a2 , but I think this specific patch is just enforcing existing limitations and this is not really adding new constraints to FW (@RanderWang had similar change in https://github.com/thesofproject/sof/commit/ec1aeca427e8aa46e3f77aec9877fbeb9d0979f7 ). 

So in the end, if the virtual sink needs to be in precise, this PR is a sensible way to add it.
I have a few review comments to topology code added in generic files.

PS Can you also do one review (in addition to author) from @thesofproject/amd to indicate the PR is ready to go?

File: tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4 (hunk 11 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof/pipe-passthrough-playback-sched.m4]: added virtual in the file name.

File: tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4 (hunk 12 above)
Comment (maruthi-machani) [reply tools/topology/topology1/sof-acp-rmb-tdm8ch-dmic2ch-nau8825-max98360.m4:1]: used topology1:amd

**Review by bhiregoudar — APPROVED** [review]

**Review by bhiregoudar — APPROVED** [review]

**Review by bhiregoudar — APPROVED** [review]

**Comment by bhiregoudar** [comment]
> @bhiregou The explanation in comments helped. It would help to have more of the context also in git commits to help the review.
> 
> This is specially complex case, so extra verbosity will help. First, I understand you are working around an issue with TDM here. I don't fully understand why ACPHS1 needs to be present in topology (if no data is written to it), but I assume there's some dependency/need for this, and this has been in use before this PR. But this definitely is not a common topology construct.
> 
> Then the problem hit here is that SOF pipeline code assumes there's a single sink and source for a pipeline. Functions like ipc_get_ppl_sink_comp() and ipc_get_ppl_src_comp() are used in generic pipeline code and they assume a single component to be returned. You hit this with [e8237f3](https://github.com/thesofproject/sof/commit/e8237f3e0b0c106983b3caa1aa4af417f08080a2) , but I think this specific patch is just enforcing existing limitations and this is not really adding new constraints to FW (@RanderWang had similar change in [ec1aeca](https://github.com/thesofproject/sof/commit/ec1aeca427e8aa46e3f77aec9877fbeb9d0979f7) ).
> 
> So in the end, if the virtual sink needs to be in precise, this PR is a sensible way to add it. I have a few review comments to topology code added in generic files.
> 
> PS Can you also do one review (in addition to author) from @thesofproject/amd to indicate the PR is ready to go?

ACPHS1 is required to show another end point is present in the system and send params IPC to fw during corresponding stream start. We are sending stream data to speaker and headset end points for simultaneous playback in TDM mode using MUX component with single I2S controller data path.

**Comment by plbossart** [comment]
> ACPHS1 is required to show another end point is present in the system

You've repeated that explanation multiple times and I still don't get the concept. As mentioned by @kv2019i this is a very unusual topology and it's really unclear if you have a legit reason for this construct, added ACPHS1 as a work-around or if the design works by accident.

So again here are the questions:

a) What agent in the stack will look for the presence of this unused hardware interface in the topology?
b) What sort of IPCs do you need to send to ACPHS1 and why?

**Comment by bhiregoudar** [comment]
> > ACPHS1 is required to show another end point is present in the system
> 
> You've repeated that explanation multiple times and I still don't get the concept. As mentioned by @kv2019i this is a very unusual topology and it's really unclear if you have a legit reason for this construct, added ACPHS1 as a work-around or if the design works by accident.

This is not design works by accident. ACPHS1 is added as a workaround to mitigate hardware limitation to support simultaneous playback on headset and speaker with one I2S controller through TDM mode.  
> 
> So again here are the questions:
> 
> a) What agent in the stack will look for the presence of this unused hardware interface in the topology? 

1) Need to show two end points(backend) present in the system by mentioning in the topology.
2) BE DAI link needs to configure the codec's TDM slots in machine driver.

>b) What sort of IPCs do you need to send to ACPHS1 and why?

configure pipeline audio params  and stream trigger. Required to configure backend codec params

**Comment by plbossart** [comment]
> 1. Need to show two end points(backend) present in the system by mentioning in the topology.

I still don't know why this is needed. We have used TDM before, we've never had such a topology with a fake branch.

I wonder if this is a result of your machine driver exposing 2 DAIs on the CPU side, you can use a single DAI on the CPU side even if you have two separate codecs attached to the TDM link. You do not need to expose a pretend-DAI at all.

File: src/platform/amd/rembrandt/include/platform/lib/memory.h [hunk 13]
```diff
@@ -98,9 +98,9 @@
 #define HEAP_RT_COUNT2048	12
 
 /* Heap section sizes for system runtime heap */
-#define HEAP_SYS_RT_COUNT64	64
-#define HEAP_SYS_RT_COUNT512	20 /*rembrandt-arch*/
-#define HEAP_SYS_RT_COUNT1024	6
+#define HEAP_SYS_RT_COUNT64	72
+#define HEAP_SYS_RT_COUNT512	24 /*rembrandt-arch*/
+#define HEAP_SYS_RT_COUNT1024	8
```
Comment (lgirdwood) [inline src/platform/amd/rembrandt/include/platform/lib/memory.h:103]: The good thing is that this is not needed when using Zephyr :)

**Review by lgirdwood — APPROVED** [review]

**Comment by wszypelt** [comment]
@lgirdwood PR is good, you can merge :)
